### ISSUE TITLE
docker ipv6 link-local address cannot be bound to for first 1-2 seconds

### ISSUE THREAD

**Issue by sethkoehler** [moby/moby#38491]
**Description**

When trying to bind to an ipv6 link-local address (fe80:...), for the first 1-2 seconds after a container starts, the bind will fail.

**Steps to reproduce the issue:**
1.  Extract [cc_net.zip](https://github.com/moby/moby/files/2725809/cc_net.zip)
2.  Run ./test.sh <seconds to sleep before binding>.  This compiles and runs inside docker a 40 line C++ program that sleeps the specified number of seconds, then goes through IPv6 interfaces on the system and tries to bind to a fixed port on each

**Describe the results you received:**
If you run ./test.sh 0, the program will fail, while ./test.sh 2 should succeed (and ./test.sh 1 will sometimes succeed, but most often fail on my system).

**Describe the results you expected:**
The binding should work for both 0 and 1 second sleeps.

**Additional information you deem important (e.g. issue happens only occasionally):**
On my machine, the program will also attempt to bind to ::1 as well, and this succeeds even when delaying for 0 seconds, so the issue doesn't appear with ::1, only the fe80:... link-local address.  Haven't tried other IPv6 addresses.

**Output of `docker version`:**

Client:
Version: 18.06.1-ce
API version: 1.38
Go version: go1.10.3
Git commit: e68fc7a
Built: Tue Aug 21 17:23:18 2018
OS/Arch: linux/amd64
Experimental: false

Server:
Engine:
Version: 18.06.1-ce
API version: 1.38 (minimum version 1.12)
Go version: go1.10.3
Git commit: e68fc7a
Built: Tue Aug 21 17:22:21 2018
OS/Arch: linux/amd64
Experimental: false

Note that I've also tried this on a few older docker versions and noticed the same, such as 17.09 and 17.12.

**Output of `docker info`:**

Containers: 30
 Running: 0
 Paused: 0
 Stopped: 30
Images: 32
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.17.0-3<redacted>-amd64
Operating System: Debian GNU/Linux buster/sid
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 31.39GiB
Name: <redacted>
ID: KGDS:45BO:QE4C:K3DI:3KSP:ORV2:RGQY:VB6W:XIA7:U4B3:6I23:QRDO
Docker Root Dir: /usr/local/<redacted>/docker/lib_docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support

**Additional environment details (AWS, VirtualBox, physical, etc.):**
I've tried this on both a physical box and on a GCE VM, with the same results.

**Comment by thaJeztah** [comment]
ping @fcrisciani ptal

**Comment by justincormack** [comment]
RFC 4862 says that IPv6 link local addresses must perform duplicate address detection before being usable, see https://tools.ietf.org/html/rfc4862#section-5.4 so you won't be able to bind until this takes place (even though there should not be conflicts in this case). I suspect this timeout is how long the kernel is taking to finish that, so I doubt there is anything we can do. You can almost certainly use netlink to wait for the address to be assigned, or retry, or use `::`.

**Comment by sethkoehler** [comment]
Thanks for the quick response, that's good to know!  I'm now wondering...is it right that the container has started and is running actions if the network that the user requested it setup (IPv6 via the --sysctl net.ipv6.conf.all.disable_ipv6=0 flag) isn't ready yet?  Should the container wait until the network has finished it's duplicate address detection (I'm assuming there's some way to detect that...you mentioned netlink, though I'm not sure whether that's for a specific address or more general) before starting the user's action?

The alternative as you mentioned is that all user actions have to be aware of the fact that the network may not be running at that moment, but may come up shortly.  My fear here is that most folks won't be writing such error-handling code that way outside of docker since the machine and network will almost inevitably be up, and thus they won't run into this issue.  Since these programs may fail inside docker due to this issue, they'll see it as a "docker issue", even though it's technically not as you point out, and users will have to figure that out and write some additional error handling to get their program "working with docker".

From my experience, going the "delay" route would delay the container starting by ~0.5-1.5 seconds (if IPv6 networking were requested), which I'd agree isn't great, but it also doesn't seem great to start running actions before user-requested resources are ready, and the delay seems cleaner from a user perspective.  What do you think?

**Comment by justincormack** [comment]
Yes, it might be possible to get libnetwork to wait (and any CNI plugins if people are using Kubernetes).

I think this will only happen with the link local address, all the other addresses should be set correctly. I am curious why you are using the link local address, it is a slightly unusual requirement...

**Comment by sethkoehler** [comment]
In this case, this is a unit test for gRPC that is verifying some basic properties across all IPv4 and IPv6 interfaces (see https://github.com/grpc/grpc/blob/1b6c3a01cf292159cd4208e8fbe64f84c8b90a8b/test/core/iomgr/tcp_server_posix_test.cc#L481).  When cycling through all interfaces, the link-local is included, since it's setup by docker, and thus the test attempts to bind a port on this address.
